improver/nowcasting/optical_flow.py

```diff
@@ -731,41 +807,46 @@ def process(self, cube1, cube2, boxsize=30):
         self.data_smoothing_radius = None
         self.boxsize = None
 
-        # check the nature of the input cubes, and raise a warning if they are
-        # not both precipitation
-        if cube1.name() != cube2.name():
-            msg = "Input cubes contain different data types {} and {}"
-            raise ValueError(msg.format(cube1.name(), cube2.name()))
+        # check input cubes have appropriate and matching contents and dimensions
+        self._check_input_cubes(cube1, cube2)
+
+        # get time difference between cubes
+        if perturbation:
+            # check the observation cube has a forecast period of zero
+            if (
+                cube2.coords("forecast_period")
+                and cube2.coord("forecast_period").points[0] != 0
+            ):
+                raise ValueError("The second input cube must be a current observation")
+            # get the time difference from the first cube's forecast period
+            fp_coord = cube1.coord("forecast_period").copy()
+            fp_coord.convert_units("seconds")
+            (time_diff_seconds,) = fp_coord.points
+        else:
+            time_diff_seconds = self._get_input_time_difference(cube1, cube2)
```

Comment:
The generalised form is 
`time_diff_seconds = (cube2.coord('time').points - cube1.coord('time').points) + (cube2.coord('forecast_reference_time').points - cube1.coord('forecast_reference_time').points)`
With this, you wouldn't need the `if perturbation` block here.

I think this would be simpler if you got rid of the `perturbation` argument and tested whether there was a difference in the `time` coord or in the `forecast_reference_time` coord (both could be done with the `_get_input_time_difference`), set `time_diff_seconds` from whichever is non-zero and positive and proceed from there. Differences on both should be rejected - there's no use-case for that yet.

Reply:
OK - will experiment.